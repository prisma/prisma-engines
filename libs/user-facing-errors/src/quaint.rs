use crate::{common, query_engine, KnownError};
use common::ModelKind;
use indoc::formatdoc;
use quaint::{error::ErrorKind, prelude::ConnectionInfo};

impl From<&quaint::error::DatabaseConstraint> for query_engine::DatabaseConstraint {
    fn from(other: &quaint::error::DatabaseConstraint) -> Self {
        match other {
            quaint::error::DatabaseConstraint::Fields(fields) => Self::Fields(fields.to_vec()),
            quaint::error::DatabaseConstraint::Index(index) => Self::Index(index.to_string()),
            quaint::error::DatabaseConstraint::ForeignKey => Self::ForeignKey,
            quaint::error::DatabaseConstraint::CannotParse => Self::CannotParse,
        }
    }
}

impl From<quaint::error::DatabaseConstraint> for query_engine::DatabaseConstraint {
    fn from(other: quaint::error::DatabaseConstraint) -> Self {
        match other {
            quaint::error::DatabaseConstraint::Fields(fields) => Self::Fields(fields.to_vec()),
            quaint::error::DatabaseConstraint::Index(index) => Self::Index(index),
            quaint::error::DatabaseConstraint::ForeignKey => Self::ForeignKey,
            quaint::error::DatabaseConstraint::CannotParse => Self::CannotParse,
        }
    }
}

pub fn invalid_connection_string_description(error_details: &str) -> String {
    let docs = r#"https://www.prisma.io/docs/reference/database-reference/connection-urls"#;

    let details = formatdoc! {r#"
            {} in database URL. Please refer to the documentation in {} for constructing a correct
            connection string. In some cases, certain characters must be escaped. Please
            check the string for any illegal characters."#, error_details, docs};

    details.replace('\n', " ")
}

pub fn render_quaint_error(kind: &ErrorKind, connection_info: &ConnectionInfo) -> Option<KnownError> {
    match (kind, connection_info) {
        (ErrorKind::DatabaseDoesNotExist { .. }, ConnectionInfo::Sqlite { file_path, .. }) => {
            Some(KnownError::new(common::DatabaseDoesNotExist::Sqlite {
                database_file_path: file_path.clone(),
                database_file_name: std::path::Path::new(file_path)
                    .file_name()
                    .map(|osstr| osstr.to_string_lossy().into_owned())
                    .unwrap_or_else(|| file_path.clone()),
            }))
        }

        (ErrorKind::DatabaseDoesNotExist { .. }, ConnectionInfo::Postgres(url)) => {
            Some(KnownError::new(common::DatabaseDoesNotExist::Postgres {
                database_name: url.dbname().to_owned(),
                database_schema_name: url.schema().to_owned(),
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::DatabaseDoesNotExist { .. }, ConnectionInfo::Mysql(url)) => {
            Some(KnownError::new(common::DatabaseDoesNotExist::Mysql {
                database_name: url.dbname().to_owned(),
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }
        (ErrorKind::DatabaseDoesNotExist { .. }, ConnectionInfo::Mssql(url)) => {
            Some(KnownError::new(common::DatabaseDoesNotExist::Mssql {
                database_name: url.dbname().to_owned(),
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::DatabaseAccessDenied { .. }, ConnectionInfo::Postgres(url)) => {
            Some(KnownError::new(common::DatabaseAccessDenied {
                database_user: url.username().into_owned(),
                database_name: format!("{}.{}", url.dbname(), url.schema()),
            }))
        }

        (ErrorKind::DatabaseAccessDenied { .. }, ConnectionInfo::Mysql(url)) => {
            Some(KnownError::new(common::DatabaseAccessDenied {
                database_user: url.username().into_owned(),
                database_name: url.dbname().to_owned(),
            }))
        }

        (ErrorKind::DatabaseAlreadyExists { db_name }, ConnectionInfo::Postgres(url)) => {
            Some(KnownError::new(common::DatabaseAlreadyExists {
                database_name: format!("{}", db_name),
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::DatabaseAlreadyExists { db_name }, ConnectionInfo::Mysql(url)) => {
            Some(KnownError::new(common::DatabaseAlreadyExists {
                database_name: format!("{}", db_name),
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::AuthenticationFailed { user }, ConnectionInfo::Postgres(url)) => {
            Some(KnownError::new(common::IncorrectDatabaseCredentials {
                database_user: format!("{}", user),
                database_host: url.host().to_owned(),
            }))
        }

        (ErrorKind::AuthenticationFailed { user }, ConnectionInfo::Mysql(url)) => {
            Some(KnownError::new(common::IncorrectDatabaseCredentials {
                database_user: format!("{}", user),
                database_host: url.host().to_owned(),
            }))
        }

        (ErrorKind::ConnectionError(_), ConnectionInfo::Postgres(url)) => {
            Some(KnownError::new(common::DatabaseNotReachable {
                database_port: url.port(),
                database_host: url.host().to_owned(),
            }))
        }

        (ErrorKind::ConnectionError(_), ConnectionInfo::Mysql(url)) => {
            Some(KnownError::new(common::DatabaseNotReachable {
                database_port: url.port(),
                database_host: url.host().to_owned(),
            }))
        }

        (ErrorKind::UniqueConstraintViolation { constraint }, _) => {
            Some(KnownError::new(query_engine::UniqueKeyViolation {
                constraint: constraint.into(),
            }))
        }

        (ErrorKind::TlsError { message }, _) => Some(KnownError::new(common::TlsConnectionError {
            message: message.into(),
        })),

        (ErrorKind::ConnectTimeout, ConnectionInfo::Mysql(url)) => {
            Some(KnownError::new(common::DatabaseNotReachable {
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::ConnectTimeout, ConnectionInfo::Postgres(url)) => {
            Some(KnownError::new(common::DatabaseNotReachable {
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::ConnectTimeout, ConnectionInfo::Mssql(url)) => {
            Some(KnownError::new(common::DatabaseNotReachable {
                database_host: url.host().to_owned(),
                database_port: url.port(),
            }))
        }

        (ErrorKind::SocketTimeout, ConnectionInfo::Mysql(url)) => {
            let time = match url.socket_timeout() {
                Some(dur) => format!("{}s", dur.as_secs()),
                None => String::from("N/A"),
            };

            Some(KnownError::new(common::DatabaseOperationTimeout {
                time,
                context: "Socket timeout (the database failed to respond to a query within the configured timeout — see https://pris.ly/d/mysql-connector for more details.)."
                    .into(),
            }))
        }

        (ErrorKind::SocketTimeout, ConnectionInfo::Postgres(url)) => {
            let time = match url.socket_timeout() {
                Some(dur) => format!("{}s", dur.as_secs()),
                None => String::from("N/A"),
            };

            Some(KnownError::new(common::DatabaseOperationTimeout {
                time,
                context: "Socket timeout (the database failed to respond to a query within the configured timeout — see https://pris.ly/d/mssql-connector for more details.)."
                    .into(),
            }))
        }

        (ErrorKind::SocketTimeout, ConnectionInfo::Mssql(url)) => {
            let time = match url.socket_timeout() {
                Some(dur) => format!("{}s", dur.as_secs()),
                None => String::from("N/A"),
            };

            Some(KnownError::new(common::DatabaseOperationTimeout {
                time,
                context: "Socket timeout (the database failed to respond to a query within the configured timeout — see https://pris.ly/d/postgres-connector for more details.)."
                    .into(),
            }))
        }

        (ErrorKind::PoolTimeout { max_open, timeout, .. }, _) => Some(KnownError::new(query_engine::PoolTimeout {
            connection_limit: *max_open,
            timeout: *timeout,
        })),

        (ErrorKind::DatabaseUrlIsInvalid(details), _connection_info) => {
            Some(KnownError::new(common::InvalidConnectionString {
                details: details.to_owned(),
            }))
        }

        (ErrorKind::LengthMismatch { column }, _connection_info) => {
            Some(KnownError::new(query_engine::InputValueTooLong {
                column_name: format!("{}", column),
            }))
        }

        (ErrorKind::ValueOutOfRange { message }, _connection_info) => {
            Some(KnownError::new(query_engine::ValueOutOfRange {
                details: message.clone(),
            }))
        }

        (ErrorKind::TableDoesNotExist { table: model }, ConnectionInfo::Mysql(_)) => {
            Some(KnownError::new(common::InvalidModel {
                model: format!("{}", model),
                kind: ModelKind::Table,
            }))
        }

        (ErrorKind::TableDoesNotExist { table: model }, ConnectionInfo::Postgres(_)) => {
            Some(KnownError::new(common::InvalidModel {
                model: format!("{}", model),
                kind: ModelKind::Table,
            }))
        }

        (ErrorKind::TableDoesNotExist { table: model }, ConnectionInfo::Sqlite { .. }) => {
            Some(KnownError::new(common::InvalidModel {
                model: format!("{}", model),
                kind: ModelKind::Table,
            }))
        }

        (ErrorKind::TableDoesNotExist { table: model }, ConnectionInfo::Mssql(_)) => {
            Some(KnownError::new(common::InvalidModel {
                model: format!("{}", model),
                kind: ModelKind::Table,
            }))
        }

        (ErrorKind::IncorrectNumberOfParameters { expected, actual }, ConnectionInfo::Mssql(_)) => {
            Some(KnownError::new(common::IncorrectNumberOfParameters {
                expected: *expected,
                actual: *actual,
            }))
        }

        (ErrorKind::ConnectionClosed, _) => Some(KnownError::new(common::ConnectionClosed)),

        _ => None,
    }
}
