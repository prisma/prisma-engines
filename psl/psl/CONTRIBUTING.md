# PSL

## Tests

### Running the tests

To run the PSL test suite, run `cargo test` from this directory.

### Contributing tests

Many tests are plain cargo test functions using the `#[test]` attribute.

There are also declarative test suites that do not involve writing Rust code:

- **validation tests** are located in the `tests/validation` directory. Each test is a single Prisma schema. The comment at the end of the file is a snapshot of the validation errors that PSL would return upon validating the schema.
- **reformatter tests** are located in the `tests/reformatter` directory. Each test is a Prisma schema. The test suite tests its reformatted version against a snapshot that is a file next to it, with the same name, except that it has the `.reformatted.prisma` extension instead of `.prisma`.

These tests are transformed into Rust code via the build script in `build.rs`.
