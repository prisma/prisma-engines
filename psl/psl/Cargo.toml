[package]
name = "psl"
version = "0.1.0"
edition = "2021"

[dependencies]
builtin-psl-connectors = { path = "../builtin-connectors" }
dml = { path = "../../libs/dml" }
psl-core = { path = "../psl-core" }

[dev-dependencies]
dissimilar = "1.0.4"
expect-test = "1.1.0"
indoc = "1"
native-types = { path = "../../libs/native-types" }

[features]
default_generators = ["dml/default_generators"]
