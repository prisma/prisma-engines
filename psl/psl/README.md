# PSL stands for Prisma Schema Language

This is the fa√ßade crate for the Prisma Schema Language implementation.
