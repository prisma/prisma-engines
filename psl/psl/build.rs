use std::{env, fs, io::Write as _, path};

const VALIDATIONS_ROOT_DIR: &str = "tests/validation";
const REFORMAT_ROOT_DIR: &str = "tests/reformatter";
const CARGO_MANIFEST_DIR: &str = env!("CARGO_MANIFEST_DIR");

fn main() {
    build_validation_tests();
    build_reformat_tests();
}

fn build_reformat_tests() {
    println!("cargo:rerun-if-changed={REFORMAT_ROOT_DIR}");

    let mut all_schemas = Vec::new();
    find_all_schemas("", &mut all_schemas, REFORMAT_ROOT_DIR);

    let mut out_file = out_file("reformat_tests.rs");
    let schemas_to_reformat = all_schemas.iter().filter(|name| !name.ends_with(".reformatted.prisma"));

    for schema_path in schemas_to_reformat {
        let test_name = test_name(schema_path);
        let file_path = schema_path.trim_start_matches('/');
        writeln!(
            out_file,
            "#[test] fn {test_name}() {{ run_reformat_test(\"{file_path}\"); }}"
        )
        .unwrap();
    }
}

fn build_validation_tests() {
    println!("cargo:rerun-if-changed={VALIDATIONS_ROOT_DIR}");

    let mut all_schemas = Vec::new();
    find_all_schemas("", &mut all_schemas, VALIDATIONS_ROOT_DIR);

    let mut out_file = out_file("validation_tests.rs");

    for schema_path in &all_schemas {
        let test_name = test_name(schema_path);
        let file_path = schema_path.trim_start_matches('/');
        writeln!(
            out_file,
            "#[test] fn {test_name}() {{ run_validation_test(\"{file_path}\"); }}"
        )
        .unwrap();
    }
}

fn find_all_schemas(prefix: &str, all_schemas: &mut Vec<String>, root_dir: &'static str) {
    for entry in fs::read_dir(format!("{CARGO_MANIFEST_DIR}/{root_dir}/{prefix}")).unwrap() {
        let entry = entry.unwrap();
        let file_name = entry.file_name();
        let file_name = file_name.to_str().unwrap();
        let entry_path = format!("{}/{}", prefix, file_name);
        let file_type = entry.file_type().unwrap();

        if file_name == "." || file_name == ".." {
            continue;
        }

        if file_type.is_file() {
            all_schemas.push(entry_path);
        } else if file_type.is_dir() {
            find_all_schemas(&entry_path, all_schemas, root_dir);
        }
    }
}

fn out_file(name: &str) -> std::fs::File {
    let out_dir = env::var("OUT_DIR").unwrap();
    let out_file_path = path::Path::new(&out_dir).join(name);
    fs::File::create(out_file_path).unwrap()
}

fn test_name(schema_file_path: &str) -> String {
    schema_file_path
        .trim_start_matches('/')
        .trim_end_matches(".prisma")
        .replace(['/', '\\'], "_")
}
