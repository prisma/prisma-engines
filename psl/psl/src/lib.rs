#![doc = include_str!("../README.md")]
#![deny(rust_2018_idioms, unsafe_code, missing_docs)]

pub use builtin_psl_connectors as builtin_connectors;
pub use dml::{self, lift};
pub use psl_core::{
    datamodel_connector,
    diagnostics::{self, Diagnostics},
    is_reserved_type_name,
    mcf::config_to_mcf_json_value as get_config,
    mcf::{generators_to_json, render_sources_to_json}, // for tests
    parser_database::{self, SourceFile},
    reformat,
    schema_ast,
    Configuration,
    Datasource,
    DatasourceConnectorData,
    Generator,
    PreviewFeature,
    RelationNames,
    StringFromEnvVar,
    ValidatedSchema,
    ALL_PREVIEW_FEATURES,
};

/// The implementation of the CLI getConfig() utility and its JSON format.
pub mod get_config {
    pub use psl_core::mcf::{config_to_mcf_json_value as get_config, *};
}

/// Parses and validate a schema, but skip analyzing everything except datasource and generator
/// blocks.
pub fn parse_configuration(schema: &str) -> Result<Configuration, Diagnostics> {
    psl_core::parse_configuration(schema, builtin_connectors::BUILTIN_CONNECTORS)
}

/// Parse and analyze a Prisma schema.
pub fn parse_schema(file: impl Into<SourceFile>) -> Result<ValidatedSchema, String> {
    let mut schema = validate(file.into());
    schema
        .diagnostics
        .to_result()
        .map_err(|err| err.to_pretty_string("schema.prisma", schema.db.source()))?;
    Ok(schema)
}

/// The most general API for dealing with Prisma schemas. It accumulates what analysis and
/// validation information it can, and returns it along with any error and warning diagnostics.
pub fn validate(file: SourceFile) -> ValidatedSchema {
    psl_core::validate(file, builtin_connectors::BUILTIN_CONNECTORS)
}
