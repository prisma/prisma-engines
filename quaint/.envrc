export TEST_MYSQL="mysql://root:prisma@localhost:3306/prisma"
export TEST_MYSQL8="mysql://root:prisma@localhost:3307/prisma"
export TEST_MYSQL_MARIADB="mysql://root:prisma@localhost:3308/prisma"
export TEST_PSQL="postgres://postgres:prisma@localhost:5432/postgres"
export TEST_CRDB="postgresql://prisma@127.0.0.1:26259/postgres"
export TEST_MSSQL="jdbc:sqlserver://localhost:1433;database=master;user=SA;password=<YourStrong@Passw0rd>;trustServerCertificate=true"
if command -v nix-shell &> /dev/null
then
    use nix
fi
