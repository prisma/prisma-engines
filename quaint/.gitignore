**/target
**/*.rs.bk
.direnv/
.vscode/

docker-compose.override.yml
db/
!db/.gitkeep
