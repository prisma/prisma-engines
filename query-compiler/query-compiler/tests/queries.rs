use std::{fs, sync::Arc};

use quaint::{
    prelude::{ConnectionInfo, ExternalConnectionInfo, SqlFamily},
    visitor::Postgres,
};
use query_core::{QueryDocument, QueryGraphBuilder};
use query_structure::psl;
use request_handlers::{JsonBody, JsonSingleQuery, RequestBody};
use sql_query_builder::{Context, SqlQueryBuilder};

#[test]
fn queries() {
    insta::glob!("data/*.json", |path| {
        let schema_string = include_str!("data/schema.prisma");
        let schema = psl::validate(schema_string.into());

        assert!(!schema.diagnostics.has_errors(), "{:?}", schema.diagnostics);

        let schema = Arc::new(schema);
        let query_schema = Arc::new(query_core::schema::build(schema, true));

        let connection_info = ConnectionInfo::External(ExternalConnectionInfo::new(
            SqlFamily::Postgres,
            "public".to_owned(),
            None,
        ));

        let query = fs::read_to_string(path).unwrap();
        let query: JsonSingleQuery = serde_json::from_str(&query).unwrap();

        let request = RequestBody::Json(JsonBody::Single(query));
        let doc = request.into_doc(&query_schema).unwrap();

        let QueryDocument::Single(query) = doc else {
            panic!("expected single query");
        };

        let (graph, _serializer) = QueryGraphBuilder::new(&query_schema).build(query).unwrap();

        let ctx = Context::new(&connection_info, None);
        let builder = SqlQueryBuilder::<Postgres<'_>>::new(ctx);

        let expr = query_compiler::translate(graph, &builder).unwrap();
        let pretty = expr.pretty_print(false, 80).unwrap();
        insta::assert_snapshot!(pretty);
    });
}
