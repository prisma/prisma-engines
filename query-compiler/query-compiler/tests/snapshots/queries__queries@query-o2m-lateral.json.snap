---
source: query-compiler/query-compiler/tests/queries.rs
expression: pretty
input_file: query-compiler/query-compiler/tests/data/query-o2m-lateral.json
---
dataMap {
    id: Int (id)
    email: String (email)
    activations: {
        issued: Date (issued)
        secret: String (secret)
        done: Boolean (done)
    }
}
query «SELECT "t0"."id", "t0"."email", "User_activations"."__prisma_data__" AS
       "activations" FROM "public"."User" AS "t0" LEFT JOIN LATERAL (SELECT
       COALESCE(JSONB_AGG("__prisma_data__"), '[]') AS "__prisma_data__" FROM
       (SELECT "t3"."__prisma_data__" FROM (SELECT JSONB_BUILD_OBJECT('issued',
       "t2"."issued", 'secret', "t2"."secret", 'done', "t2"."done") AS
       "__prisma_data__" FROM (SELECT "t1".* FROM "public"."Activation" AS "t1"
       WHERE "t0"."id" = "t1"."userId" /* root select */) AS "t2" /* inner
       select */) AS "t3" /* middle select */) AS "t4" /* outer select */) AS
       "User_activations" ON true»
params []
